### PR TITLE
PiLog Tests

### LINKED ISSUE #110: Add tests for PiLog
A `PiLogProof` has seven elements that must satisfy four constraints. (Note this corresponds to Fig 25, `Π^log*`).
Some ideas for tests:

- Swap each of the proof elements for a different, random element (of the correct type). Proof shouldn't verify.
- Try making the proof with inconsistent secret inputs (e.g. `x` is not used to make only one of `C` and `X`, `rho` is not the correct nonce for `C`). Proof either shouldn't build correctly or shouldn't verify.
- Try making the proof with consistent secret inputs that are out of range (e.g. `x` is actually used to make `C` and `X`, but it's larger than `2^ell`).
- Try verifying the proof with different setup parameters than were used to construct it.

Related issue: #48

### PR DESCRIPTION
**Pull request by hridambasu**
Closes #110

### CODE REVIEW

File: src/zkp/pilog.rs [hunk 1]
```diff
@@ -393,6 +393,147 @@ mod tests {
...
+        let scheme = setup_params.scheme();
+        let (bad_plaintext_mask, bad_randomness) = scheme.commit(&random_mask, ELL, &mut rng);
+
+        // Swap mask_commit with a random [`Commitment`]
+        let mut bad_proof = proof.clone();
+        bad_proof.mask_commit = bad_plaintext_mask.clone();
+        assert_ne!(bad_proof.mask_commit, proof.mask_commit);
+        assert!(bad_proof.verify(&input, &(), &mut transcript).is_err());
+
+        // Swap mask_commit with a random [`Commitment`]
```
Comment (marsella) [inline src/zkp/pilog.rs:481]: nit: fix comment.

File: src/zkp/pilog.rs [hunk 2]
```diff
@@ -393,6 +393,147 @@ mod tests {
...
+            &(),
+            &mut transcript,
+            &mut rng,
+        )?;
+        assert!(bad_proof.verify(&input, &(), &mut transcript).is_err());
+        Ok(())
+    }
+
+    #[test]
+    fn negative_test_swap_proof_elements() -> Result<()> {
```
Comment (marsella) [inline src/zkp/pilog.rs:463]: This test looks great!

**Review by marsella — CHANGES_REQUESTED** [review]
Thank you, this is a good start! These tests can be quite fiddly so many of my comments are naming / documentation tweaks. I also think we're missing a couple of tests:
- the negative side of the range proof
- the bad-rho half of the "you need to know the secret input" test
- the whole common-input-must-match test.

File: src/zkp/pilog.rs [hunk 3]
```diff
@@ -393,6 +393,147 @@ mod tests {
...
+                random_paillier_log_proof(&mut rng, &too_large).unwrap();
+            if too_large.gt(&upper_bound) {
+                assert!(bad_proof.verify(&input, &(), &mut transcript).is_err());
+                break;
+            }
+        }
+        /*let too_small = -too_large;
+        let (bad_proof, input, mut transcript) =
+            random_paillier_log_proof(&mut rng, &too_small).unwrap();
+        assert!(bad_proof.verify(&input, &(), &mut transcript).is_err());*/
```
[the remark this answers is not in the document]
Comment (marsella) [reply src/zkp/pilog.rs]: Oops, never mind, I started reviewing too soon.

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks, this looks better! I have a few more refactoring / renaming kinds of comments.

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks, this is looking much better. Only one small comment on your work but I realized two things:
1. Transcripts are not being initialized correctly throughout -- we should use a better paradigm to make sure each prove/verify call gets a fresh transcript (see comment inline)
2. We had different ideas about what the setup parameters test should be testing. Would like your thoughts on what the meaningful failure conditions of the proofs are (and how that informs which tests we should write).

**Comment by hridambasu** [comment]
> 2. We had different ideas about what the setup parameters test should be testing. Would like your thoughts on what the meaningful failure conditions of the proofs are (and how that informs which tests we should write).

I can add another test to mimic the exact sequence of events in the different setup parameters test by creating a valid proof first and then asserting that the proof fails to verify if one of the input fields are modified. But honestly after the test that I added, it seems a bit redundant. But I can also add that test in addition to my test or modify my test with the one you suggested.

File: src/zkp/pilog.rs [hunk 4]
```diff
@@ -349,9 +349,11 @@ impl Proof for PiLogProof {
 
 #[cfg(test)]
 mod tests {
+    //use k256::elliptic_curve::bigint::Random;
```
Comment (marsella) [inline src/zkp/pilog.rs:352]: todo: remove commented-out code.

**Review by marsella — APPROVED** [review]
We talked about the common-input test offline -- the tests would test separate things: this one checks that you can't create a proof that passes if your input is wrong, and the one I suggested checks that you can't verify a valid proof if you have the wrong input. We decided that the latter is not a high priority test for this setting because of how it could happen (either something gets messed up in transit, or a malicious prover lies to the verifier and causes their own proof to fail, or a malicious verifier takes an extremely roundabout way to just abort), so we'll skip it.

Otherwise this looks good!

**Pull request merged**